app/src/main/java/com/humberto/tasky/agenda/presentation/agenda_details/AgendaDetailsViewModel.kt

```diff
@@ -2,29 +2,84 @@ package com.humberto.tasky.agenda.presentation.agenda_details
+import com.humberto.tasky.agenda.presentation.agenda_details.mapper.toAgendaDetailsUi
+import com.humberto.tasky.agenda.presentation.agenda_details.model.AgendaDetailsUi
+import com.humberto.tasky.core.domain.util.Result
+import com.humberto.tasky.core.domain.util.map
+import com.humberto.tasky.core.presentation.ui.asUiText
+import com.humberto.tasky.event.domain.EventRepository
 import com.humberto.tasky.main.navigation.AgendaDetails
+import com.humberto.tasky.reminder.domain.ReminderRepository
+import com.humberto.tasky.task.domain.TaskRepository
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.Channel
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
 import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
 import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
+@HiltViewModel
 class AgendaDetailsViewModel @Inject constructor(
-    savedStateHandle: SavedStateHandle
+    savedStateHandle: SavedStateHandle,
+    private val taskRepository: TaskRepository,
+    private val eventRepository: EventRepository,
+    private val reminderRepository: ReminderRepository
 ): ViewModel() {
 
     private val agendaDetailsArgs = savedStateHandle.toRoute<AgendaDetails>()
 
     private val _state = MutableStateFlow(
         AgendaDetailsState(
-            agendaItemType = agendaDetailsArgs.agendaItemType,
-            agendaItemId = agendaDetailsArgs.agendaItemId,
+            agendaItem = AgendaDetailsUi(
+                id = agendaDetailsArgs.agendaItemId,
+                agendaItemType = agendaDetailsArgs.agendaItemType,
+            ),
             isEditing = agendaDetailsArgs.isEditing
         )
     )
     val state: StateFlow<AgendaDetailsState> = _state.asStateFlow()
 
+    private val eventChannel = Channel<AgendaDetailsEvent>()
+    val events = eventChannel.receiveAsFlow()
+
+    init {
+        if (agendaDetailsArgs.agendaItemId != null) {
+            getItemById(
+                id = agendaDetailsArgs.agendaItemId,
+                type = agendaDetailsArgs.agendaItemType
+            )
+        }
+    }
+
+    private fun getItemById(id: String, type: AgendaItemType) {
+        viewModelScope.launch {
+            val result = when (type) {
+                AgendaItemType.TASK -> taskRepository.getTask(id).map { AgendaItem.TaskItem(it) }
+                AgendaItemType.EVENT -> eventRepository.getEvent(id).map { AgendaItem.EventItem(it) }
+                AgendaItemType.REMINDER -> reminderRepository.getReminder(id).map { AgendaItem.ReminderItem(it) }
+            }
+
+            when (result) {
+                is Result.Success -> {
+                    val agendaItem = result.data
+                    _state.value = _state.value.copy(
+                        agendaItem = agendaItem.toAgendaDetailsUi()
+                    )
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue, but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.
Using `_state.value = _state.value.copy(...)` is fine if you always call this from a single thread (which you do here). But in ViewModels, you're very likely to not do that as soon as you'd change the state inside a `viewModelScope`.